### ISSUE TITLE
Compile times are very slow - Version without type level computations possible?

### ISSUE THREAD

**Issue by minad** [Quid2/flat#8]
Hi,

I am interested in using flat, since it performs very well according to the haskell-perf benchmarks. However the compile times are very slow. I hate to say it - but would it make sense to create a less safe, but more practical version without the type level computations? I also had issues with enums with more than 255 constructors, since there is some hardcoded limit. I haven't looked into it but are the typelevel Nat computations necessary for the instance generation to get this bitwise packing?

**Comment by tittoassini** [comment]
Hi Daniel,

unfortunately, regarding the slow compilation times, I am not sure that there is much space for improvement.

Compiling a long enumeration takes about 7 seconds on my oldish macbook pro:

data E = N1 | N2 | N3 .. | N256 deriving (Generic,Flat)

Problem is: the Flat instance generation time is only about 20% of this time.

Most of the compilation time (70%) is due to the Generic instance.

I also tested with Binary, you get the same disappointing compilation times.


As for the 255 constructors limitation, it's an arbitrary limit and it could be removed.

Just out of curiosity, what is the use case for really long enumerations? Is some code that you generate?

best

**Comment by minad** [comment]
@tittoassini Thank you for your answer! Actually I tried binary, cereal and flat and only flat leads to those long compile times. I am using cereal right now. My project takes already very long due to other th and generic instances. However with flat the compilation at least doubles again (but I didn't measure that).

You say it is not due to the generic generation, but due to the compilation of the instances. Is this due to the more complex code needed for flat to achieve bitwise packing? Maybe the compiler has a hard time to simplify complex generics code? While uglier, one might consider using TH instead to generate the instances to take away a bit of the stress on the later compiler stages. In general I think the bad scaling of enums should be possible to fix somehow. I don't see a particular reason why enum instance compilation should take longer than the compilation of an integer with fixed bit width. But maybe one cannot use generics to achieve that :/

The long enumerations (~ 400 constructors) are used for a list of primitive operations in an interpreter, which is imho an acceptable use case. In particular since enums boil down to integers in the end. As a fix I wrote my own instance based on the Word16 instance. But that still left me with the long compile times for all the rest. 

In contrast to enums, I don't think there are many uses cases where one needs adts with many complex constructors which would probably lead to even more stress on the generic instance generation.

**Comment by tittoassini** [comment]
Encoding code is also more complex.

Flat first calculates the maximum size of the encoded value then allocates statically a buffer and finally encodes the value.

This is efficient at run time but means that we need to generate two methods "size" and "encode" rather than just "encode" and therefore might lead to longer compilation times.

**Comment by minad** [comment]
My concern is not only enumerations but slow compilation in general. But I still wonder why the code generated for enumerations is long as you say? Can it not be done simply by converting the enum to an integer and serializing that? Does the resulting code look like a huge case where every case alternative serializes one value? If that's the case, that is clearly suboptimal also due to the amount of generated machine code unnecessarily clogging the cpu cache etc. But you are right that in the general case there is not too much that can be done since there every constructor is indeed a *special* case which needs its own code. And then flat will take more time than binary for sure due to its encoding and size/encode functions. 

I am already keeping the datatypes in special modules, which do not contain functions. That's certainly better, but unnecessary long compilation is still an issue that I would want to avoid if possible.

**Comment by tittoassini** [comment]
Strictly speaking, Flat doesn't generate anything. 

The encoding/decoding code is just generic code operating on the base of the data type representation produced by Generic.

It's when the code is compiled that GHC expands it to be able to optimise it. 

Turning off all optimisations might remove this overhead, at least during development.

I will check and will let you know.

**Comment by tittoassini** [comment]
In principle, it would be possible to identify and treat enumerations in a more efficient way.

In fact, this might well already happen as GHC optimises the code.

The only way of knowing is looking at the final generated core code, that often has little resemblance to the source code, in particular for special mechanisms like Generics.

In fact, Flat is quite competitive with Binary and company when it comes to execution times, even if its code is more complex.

But the price for these clever optimisations is longer compilation times.

**Comment by tittoassini** [comment]
I am looking into this, I have already made a few changes that might help.

Can you try to recompile using the current version in github?

```
- location:
     git: https://github.com/Quid2/flat
     commit: 4795b519e2bd58127044d54b69dc371018607366
  extra-dep: true
```

and see if it improves things somewhat?

What I usually do is to turn optimisations off while developing (stack --fast or -0 in cabal), this totally removes this compilation time overhead.

**Comment by tysonzero** [comment]
@tittoassini 

That improved things substantially, to 114/144s using 3.5/5.5GB of RAM for ghc/ghcjs respectively.

Great work! That might be enough to merge it in, although I will have to see if we have anyone running 8GB of ram before doing so.

The above is with `O2`.

With `O0` it takes fairly trivial resources to compile (couple seconds maybe), and `O1` performs pretty similarly to `O2`. Unfortunately dropping to `O0` is probably not an option due to how performance sensitive a realtime game in JS is.

**Comment by tittoassini** [comment]
I am happy to hear that, I am making a few more changes that should improve the situation further, I will keep you posted.

**Comment by tittoassini** [comment]
I have just released flat-0.3.4, that should have significantly better compilation times than 0.3.2.

More work could be done on this but I don't think that I will have time to focus on it for a while. 


@minad The new release also increases the number of supported constructors to 512, that, as Bill Gates would say, ought to be enough for anybody.

**Comment by minad** [comment]
Cool, thank you for your work! I will test flat on my project and compare it with persist, which I made recently.
